File: src/Showdown.php
<?php

namespace App;

use App\Players\AIPlayer;
use App\Players\HumanPlayer;

class Showdown
{
    private array $players;

    private int $round = 1;

    protected array $exchangeHands;

    public function __construct(
        private readonly Deck $deck
    ) {
        $this->exchangeHands = [];
    }

    public function startGame(int $numRealPlayers, int $numAIPlayers): void
    {
        // Total players
        $totalPlayers = $numRealPlayers + $numAIPlayers;

        $playerNum = 1;
        for ($i = 0; $i < $numRealPlayers; $i++) {
            $player = new HumanPlayer($this);
            echo "請輸入玩家{$playerNum}名稱：";
            $player->nameHimself();
            $this->addPlayer($player);
            $playerNum++;
        }

        if ($numAIPlayers != 0) {
            for ($i = 0; $i < $numAIPlayers; $i++) {
                $player = new AIPlayer($this);
                $player->nameHimself();
                $this->addPlayer($player);
            }
        }

        for ($i = 0; $i < $totalPlayers; $i++) {
            $num = $i + 1;
            echo "玩家{$num}名稱為：".$this->players[$i]->getName()."\n";
        }

        echo "遊戲開始, 玩家總人數為: $totalPlayers\n";

        echo "開始洗牌\n";
        $this->deck->shuffle();
        echo "開始抽牌\n";

        // Draw 13 cards for each player
        for ($i = 0; $i < 13; $i++) {
            foreach ($this->players as $player) {
                /** @var HumanPlayer|AIPlayer $player */
                $player->drawCard($this->deck);
            }
        }

        $this->startRound();
    }

    private function addPlayer(HumanPlayer|AIPlayer $player): void
    {
        $this->players[] = $player;
    }

    public function startRound(): void
    {
        echo "第{$this->round}回合開始\n";

        // Each player plays to choose an action
        foreach ($this->players as $player) {
            /** @var HumanPlayer|AIPlayer $player */
            $player->chooseAction();
        }

        $this->compareCards();
        $this->endRound();
    }

    public function compareCards(): void
    {
        $maxIndex = 0;

        foreach ($this->players as $index => $player) {
            /** @var HumanPlayer|AIPlayer $player */
            $card = $player->getShowedCard();

            if (empty($max)) {
                $max = $card;
            }

            // Compare card rank and suit
            if ($card->getRank()->value > $max->getRank()->value) {
                $max = $card;
                $maxIndex = $index;
            } elseif ($card->getRank()->value == $max->getRank()->value) {
                if ($card->getSuit()->value > $max->getSuit()->value) {
                    $max = $card;
                    $maxIndex = $index;
                }
            }
        }

        if (empty($max)) {
            echo "本回合無人出牌\n";
        } else {
            $suit = $max->getSuit()->toCardString();
            $rank = $max->getRank()->toCardString();

            /** @var HumanPlayer|AIPlayer $winner */
            $winner = $this->players[$maxIndex];
            $winner->gainPoint();

            echo "本回合贏家為：{$winner->getName()}, 目前得分為：{$winner->getPoint()}  ";
            echo "最大的牌為：$suit $rank\n";
        }
    }

    public function endRound(): void
    {
        // Countdown exchange hands
        $exchangeHands = $this->getExchangeHands();
        if (count($exchangeHands) !== 0) {
            foreach ($exchangeHands as $exchangeHand) {
                /** @var ExchangeHands $exchangeHand */
                $exchangeHand->countdown();
                if ($exchangeHand->getDuration() === 0) {
                    $this->removeExchangeHands($exchangeHand);
                }
            }
        }

        echo "第{$this->round}回合結束\n";
        $this->round++;
        $this->round > 13 ? $this->endGame() : $this->startRound();
    }

    private function endGame(): void
    {
        $winners = $this->compareScoreWithPlayers($this->players);

        echo "遊戲結束, 贏家為：\n";
        foreach ($winners as $winner) {
            echo "{$winner->getName()}, 得分為：{$winner->getPoint()} \n";
        }
    }

    public function compareScoreWithPlayers(array $players): array
    {
        $players = array_map(fn ($player) => $player->getPoint(), $players);
        $max = max($players);

        // Winner may be more than one
        return array_filter($this->players,
            fn ($player) => $player->getPoint() === $max);
    }

    public function getPlayers(): array
    {
        return $this->players;
    }

    public function getExchangeHands(): array
    {
        return $this->exchangeHands;
    }

    public function addExchangeHands(ExchangeHands $exchangeHands): void
    {
        $this->exchangeHands[] = $exchangeHands;
    }

    public function removeExchangeHands(ExchangeHands $exchangeHands): void
    {
        $filtered = array_filter($this->exchangeHands,
            fn ($exchangeHand) => $exchangeHand !== $exchangeHands);
        $this->exchangeHands = array_values($filtered);
    }
}

File: showdown-go/entity/showdown.go
package entity

import (
	players "showdown-go/entity/player"
)

type Showdown struct {
	players       []*players.AbstractPlayer
	round         int
	exchangeHands []*ExchangeHands
	deck          *Deck
}

func NewShowdown(deck *Deck) *Showdown {
	return &Showdown{
		deck:          deck,
		exchangeHands: []*ExchangeHands{},
	}
}

func (s *Showdown) StartGame(numRealPlayers int, numAIPlayers int) {
	// Total players
	totalPlayers := numRealPlayers + numAIPlayers

	playerNum := 1
	for i := 0; i < numRealPlayers; i++ {
		player := players.NewHumanPlayer(s)
		player.NameHimself()
		s.AddPlayer(player)
		playerNum++
	}

	if numAIPlayers != 0 {
		for i := 0; i < numAIPlayers; i++ {
			player := players.NewAIPlayer(s)
			player.NameHimself()
			s.AddPlayer(player)
		}
	}

	for i := 0; i < totalPlayers; i++ {
		num := i + 1
		println("玩家", num, "名稱為：", s.players[i].GetName())
	}

	println("遊戲開始, 玩家總人數為: ", totalPlayers)

	println("開始洗牌")
	s.deck.Shuffle()
	println("開始抽牌")

	// Draw 13 cards for each player
	for i := 0; i < 13; i++ {
		for _, player := range s.players {
			player.DrawCard(s.deck)
		}
	}

	s.StartRound()
}

func (s *Showdown) AddPlayer(player *player.Player) {
	s.players = append(s.players, player)
}

func (s *Showdown) StartRound() {

	println("第", s.round, "回合開始")

	// Each player plays to choose an action
	for _, player := range s.players {
		player.ChooseAction()
	}

	s.CompareCards()
	s.EndRound()
}

func (s *Showdown) CompareCards() {
	var maximum *Card
	var maxIndex int

	for index, player := range s.players {
		card := player.GetShowedCard()

		if maximum == nil {
			maximum = card
		}

		// Compare card rank and suit
		if card.GetRank().Value > maximum.GetRank().Value {
			maximum = card
			maxIndex = index
		} else if card.GetRank().Value == maximum.GetRank().Value {
			if card.GetSuit().Value > maximum.GetSuit().Value {
				maximum = card
				maxIndex = index
			}
		}
	}

	if maximum == nil {
		println("本回合無人出牌")
	} else {
		suit := maximum.GetSuit().ToCardString()
		rank := maximum.GetRank().ToCardString()

		// Winner may be more than one
		winners := s.CompareScoreWithPlayers(s.players)

		var winnerNames []string
		var winnerPoints []int
		for _, winner := range winners {
			winnerNames = append(winnerNames, winner.GetName())
			winnerPoints = append(winnerPoints, winner.GetPoint())
		}

		println("本回合贏家為：", winnerNames, ", 目前得分為：", winnerPoints)
		println("最大的牌為：", suit, rank, "\n")
	}
}

func (s *Showdown) EndRound() {
	// Countdown exchange hands
	exchangeHands := s.GetExchangeHands()
	if len(exchangeHands) != 0 {
		for _, exchangeHand := range exchangeHands {
			exchangeHand.Countdown()
			if exchangeHand.GetDuration() == 0 {
				s.RemoveExchangeHands(exchangeHand)
			}
		}
	}

	println("第{$s.round}回合結束")
	s.round++
	if s.round > 13 {
		s.EndGame()
	} else {
		s.StartRound()
	}
}

func (s *Showdown) EndGame() {
	winners := s.CompareScoreWithPlayers(s.players)

	println("遊戲結束, 贏家為：")
	for _, winner := range winners {
		println("{$winner->GetName()}, 得分為：{$winner->GetPoint()} \n")
	}
}

func (s *Showdown) CompareScoreWithPlayers(players []*player.Player) []*player.Player {
	var playerPoints []int
	for _, player := range players {
		playerPoints = append(playerPoints, player.GetPoint())
	}

	// Winner may be more than one
	var winners []*player.Player
	for _, player := range players {
		if player.GetPoint() == max(playerPoints) {
			winners = append(winners, player)
		}
	}

	return winners
}

func (s *Showdown) GetPlayers() []*player.Player {
	return s.players
}

func (s *Showdown) GetExchangeHands() []ExchangeHands {
	return s.exchangeHands
}

func (s *Showdown) AddExchangeHands(exchangeHands *ExchangeHands) {
	s.exchangeHands = append(s.exchangeHands, exchangeHands)
}

func (s *Showdown) RemoveExchangeHands(exchangeHands *ExchangeHands) {
	var filtered []*ExchangeHands
	for _, hand := range s.exchangeHands {
		if hand != exchangeHands {
			filtered = append(filtered, hand)
		}
	}
	s.exchangeHands = filtered
}
